docs/standard/io/pipelines.md

```diff
@@ -0,0 +1,339 @@
+
+In the first loop:
+
+* <xref:System.IO.Pipelines.PipeWriter.GetMemory(System.Int32)?displayProperty=nameWithType> is called to get memory from the underlying writer.
+* <xref:System.IO.Pipelines.PipeWriter.Advance(System.Int32)?displayProperty=nameWithType>
+is called to tell the `PipeWriter` how much data was written to the buffer.
+* <xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A?displayProperty=nameWithType> is called to make the data available to the `PipeReader`.
+
+In the second loop, the `PipeReader` consumes the buffers written by `PipeWriter`. The buffers come from the socket. The call to `PipeReader.ReadAsync`:
+
+* Returns a <xref:System.IO.Pipelines.ReadResult> that contains two important pieces of information:
+
+  * The data that was read in the form of `ReadOnlySequence<byte>`.
+  * A boolean `IsCompleted` that indicates if the end of data (EOF) has been reached. 
+
+After finding the end of line (EOL) delimiter and parsing the line:
+
+* The logic processes the buffer to skip what's already processed.
+* `PipeReader.AdvanceTo` is called to tell the `PipeReader` how much data has been consumed and examined.
+
+The reader and writer loops end by calling `Complete`. `Complete` lets the underlying Pipe release the memory it allocated.
+
+### Backpressure and flow control
+
+Ideally, reading and parsing work together:
+
+* The writing thread consumes data from the network and puts it in buffers.
+* The parsing thread is responsible for constructing the appropriate data structures.
+
+Typically, parsing takes more time than just copying blocks of data from the network:
+
+* The reading thread gets ahead of the parsing thread.
+* The reading thread has to either slow down or allocate more memory to store the data for the parsing thread.
+
+For optimal performance, there's a balance between frequent pauses and allocating more memory.
+
+To solve the preceding problem, the `Pipe` has two settings to control the flow of data:
+
+* <xref:System.IO.Pipelines.PipeOptions.PauseWriterThreshold>: Determines how much data should be buffered before calls to <xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A> pause.
+* <xref:System.IO.Pipelines.PipeOptions.ResumeWriterThreshold>: Determines how much data the reader has to observe before calls to `PipeWriter.FlushAsync` resume.
+
+![Diagram with ResumeWriterThreshold and PauseWriterThreshold](./media/pipelines/resume-pause.png)
+
+<xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A?displayProperty=nameWithType>:
+
+* Returns an incomplete `ValueTask<FlushResult>` when the amount of data in the `Pipe` crosses `PauseWriterThreshold`.
+* Completes `ValueTask<FlushResult>` when it becomes lower than `ResumeWriterThreshold`.
+
+Two values are used to prevent rapid cycling, which can occur if one value is used.
```

Comment:
I'd give an example of the rapid cycling if the same value is used for the threshold.

Reply:
@jkotalik  please suggest verbiage.